shared/hooks/useDateFilter.ts

```diff
@@ -0,0 +1,61 @@
+import { DateTime } from 'luxon'
+import { useState, useEffect } from 'react'
+
+import { EventModel } from 'shared/api'
+
+type UseDateFilterReturn = {
+  fromDate: DateTime | null
+  setFromDate: (fromDate: DateTime | null) => void
+  toDate: DateTime | null
+  setToDate: (toDate: DateTime | null) => void
+  filteredEvents: EventModel[]
+  fromDateError: string | null
+  toDateError: string | null
+}
+
+const useDateFilter = (events: EventModel[] | null, translation: (key: string) => string): UseDateFilterReturn => {
+  const [fromDate, setFromDate] = useState<DateTime | null>(DateTime.local())
+  const [toDate, setToDate] = useState<DateTime | null>(DateTime.local().plus({ day: 10 }))
+  const [filteredEvents, setFilteredEvents] = useState<EventModel[]>([])
+  const [fromDateError, setFromDateError] = useState<string | null>(null)
+  const [toDateError, setToDateError] = useState<string | null>(null)
```

Comment:
These should not go to a `useState` as they are always just calculated from other states and never otherwise. Instead, you can just calculate those on the fly or, if the calculation is resource intensive, make use of `useMemo` which returns these values.

Reply:
There is a `useEffect` that prevent recalculations and it uses [fromDate, toDate, events] as dependences . 
Does it really needs useMemo ?
Oh fromDate and toDate are type `DateTime` and this is an object meaning every time a new object is passed to this dependency.(I think it called shallow comparison)
 So do I use `JSON.stringify(fromDate)` as dependency? 🙂

Reply:
I think the `stringify` is not needed. The date only changes if we reset it for some reason, in which case some of the properties usually changed.

> There is a useEffect that prevent recalculations and it uses [fromDate, toDate, events] as dependences .
Does it really needs useMemo ?

I think the question should be the other way round. Do we really need three `useState` and one `useEffect` over a simple `useMemo`? Basically, you are implementing a custom `useMemo` with that which complicates the code a bit and makes it unclear what is actually happening. Take for example the following simple `useMemo`:
```ts
  const toDateError = toDate ? null : 'invalidToDate'
  const isEarlierError = fromDate && toDate && fromDate > toDate ? 'shouldBeEarlier' : null
  const fromDateError = fromDate ? isEarlierError : 'invalidToDate'

  const filteredEvents = useMemo(() => {
    if (!fromDate || !toDate || fromDate > toDate) {
      return []
    }
    return events?.filter(event => event.date.startDate >= fromDate && event.date.endDate <= toDate.endOf('day')) || []
  }, [events, fromDate, toDate])
```

Reply:
👀 Perfect ✨ and Informative